Remotion/Web/Core/ContentSecurityPolicy/CspEnabledHtmlTextWriter.cs

```diff
@@ -0,0 +1,119 @@
+using System.Security.Cryptography;
+using System.Web.UI;
+using Remotion.Web.UI;
+
+namespace Remotion.Web.ContentSecurityPolicy
+{
+  public interface INonceGenerator
+  {
+    string GenerateAlphaNumericNonce ();
+  }
+
+  public class NonceGenerator : INonceGenerator
+  {
+    public string GenerateAlphaNumericNonce ()
+    {
+      Span<byte> data = stackalloc byte[8];
+      using (var randomNumberGenerator = RandomNumberGenerator.Create())
+      {
+        randomNumberGenerator.GetBytes(data);
+      }
+
+      return Convert.ToHexString(data);
+    }
+  }
+
+  public class CspEnabledHtmlTextWriter : HtmlTextWriter
+  {
+    private static readonly Dictionary<string, string> s_supportedEvents =
+        new(StringComparer.OrdinalIgnoreCase)
+        {
+            { "onclick", "click" },
+            { "onchange", "change" },
+            { "onmouseover", "mouseover" },
+            { "onmouseout", "mouseout" },
+            { "onkeyup", "keyup" },
+            { "onkeydown", "keydown" },
+            { "onkeypress", "keypress" }
+        };
+
+    private readonly List<Tuple<string, string>> _registeredEvents = new();
+    private readonly INonceGenerator _nonceGenerator;
+    private readonly ISmartPage _page;
+    private readonly string _nonceValue;
+
+    public CspEnabledHtmlTextWriter (ISmartPage page, TextWriter writer, string nonceValue, INonceGenerator nonceGenerator)
+        : base(writer)
+    {
+      _page = page;
+      _nonceValue = nonceValue;
+      _nonceGenerator = nonceGenerator;
+    }
+
+    public override void RenderBeginTag (HtmlTextWriterTag tagKey)
+    {
+      if (tagKey == HtmlTextWriterTag.Script)
+      {
+        AddAttribute("nonce", _nonceValue);
+      }
+
+      if (_registeredEvents.Count != 0)
+      {
+        var eventTargetID = _nonceGenerator.GenerateAlphaNumericNonce();
+        AddAttribute("data-inline-event-target", eventTargetID);
+
+        foreach (var registeredEvent in _registeredEvents)
+        {
+          var script = $"SmartPageContext.Instance.RegisterInlineEvent(\"{eventTargetID}\", \"{registeredEvent.Item1}\", function (event){{{registeredEvent.Item2}}});";
+          _page.ClientScript.RegisterStartupScriptBlock(_page, typeof(CspEnabledHtmlTextWriter), eventTargetID, script);
+        }
+      }
+
+      base.RenderBeginTag(tagKey);
+    }
+
+    public override void AddAttribute (string name, string? value)
+    {
+      if (name.StartsWith("on", StringComparison.OrdinalIgnoreCase) && !string.IsNullOrEmpty(value))
+      {
+        if (s_supportedEvents.TryGetValue(name, out var eventType))
+        {
+          if (_registeredEvents.Exists(e => e.Item1 == eventType))
+          {
+            throw new NotSupportedException($"{name} cannot be added more than once.");
+          }
+
+          _registeredEvents.Add(new Tuple<string, string>(eventType, value));
+        }
+        else
+        {
+          throw new NotSupportedException($"{name} is not supported.");
```

Comment:
ArgumentException with message 
`The name of attribute '{name}' indicates a script event but the event type is not supported by `CspEnabledHtmlTextWriter`. 
pass name and value in the arugment exception ctor.